internal/clientconn/conninfo/conn_info.go

```diff
@@ -12,28 +12,142 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package conninfo provides a ConnInfo struct that is used to handle connection-specificinfo
-// and can be shared through context.
+// Package conninfo provides access to connection-specific information.
 package conninfo
 
 import (
 	"context"
+	"math/rand"
+	"runtime"
+	"runtime/pprof"
 	"sync"
+	"sync/atomic"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/debugbuild"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
 )
 
 // contextKey is a special type to represent context.WithValue keys a bit more safely.
 type contextKey struct{}
 
-// connInfoKey stores the key for withConnInfo context value.
-var connInfoKey = contextKey{}
+var (
+	// Context key for WithConnInfo/Get.
+	connInfoKey = contextKey{}
+
+	// Keeps track on all ConnInfo objects.
+	connInfoProfiles = pprof.NewProfile("github.com/FerretDB/FerretDB/internal/clientconn/conninfo.connInfo")
+
+	// Global last cursor ID.
+	lastCursorID atomic.Uint32
+)
+
+func init() {
+	// to make debugging easier
+	if !debugbuild.Enabled {
+		lastCursorID.Store(rand.Uint32())
+	}
+}
 
 // ConnInfo represents connection info.
 type ConnInfo struct {
 	PeerAddr string
 
 	rw       sync.RWMutex
+	cursors  map[int64]Cursor
```

Comment:
Don't you wan't to use `uint32` as the map key?

Reply:
It is int64 in the wire protocol. While we could save a few bytes per cursor, that would make more code depend on the ID generation. For example, consider this code:

```
// Cursor returns cursor by ID, or nil.
func (connInfo *ConnInfo) Cursor(id int64) *Cursor {
	connInfo.rw.RLock()
	defer connInfo.rw.RUnlock()

	c, ok := connInfo.cursors[uint32(id)]
	if !ok {
		return nil
	}

	return &c
}
```

It is subtly wrong: there are two int64 values that correspond to the same uint32 value. So we have to check that id > 0 and <= math.MaxUint32. That would split that logic into two functions. Not sure if it is worth it

Reply:
ok, makes sense!